src/lem/constrainer.rs

```diff
@@ -81,103 +150,516 @@ impl LEM {
+    /// to ensure only concrete paths are implied. Otherwise, only one path
+    /// exists, therefore it must be enforced. Hence we use two distinct data
+    /// vectors, one for implications and another one for enforcements.
+    ///
+    /// Beyond that, we need to deal with the possibility of receiving no tag,
+    /// since Unhash operations don't require constraining tags.
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        alloc_ptrs: &HashMap<&String, AllocatedPtr<F>>,
+        implies_data: &Vec<(Boolean, usize, MetaPtr, Option<Tag>)>,
+        enforce_data: &Vec<(usize, MetaPtr, Option<Tag>)>,
+        hash_slots: &HashMap<usize, AllocatedNum<F>>,
+        alloc_manager: &mut AllocationManager<F>,
+    ) -> Result<()> {
+        // Create hash implications
+        for (concrete_path, slot, tgt, tag) in implies_data {
+            // get alloc_tgt from tgt
+            let Some(alloc_tgt) = alloc_ptrs.get(tgt.name()) else {
+                bail!("{} not allocated", tgt.name());
+            };
+
+            // get slot_hash from slot name
+            let Some(slot_hash) = hash_slots.get(slot) else {
+                bail!("Slot {} not allocated", slot)
+            };
+
+            implies_equal(
+                &mut cs
+                    .namespace(|| format!("implies equal hash2 for {} and {}", slot, tgt.name())),
+                concrete_path,
+                alloc_tgt.hash(),
+                slot_hash,
+            )?;
+
+            match tag {
+                Some(tag) => {
+                    let alloc_tag = alloc_manager.get_or_alloc_num(cs, tag.to_field())?;
+                    implies_equal(
+                        &mut cs.namespace(|| {
+                            format!("implies equal tag for {} and {} in hash2", slot, tgt.name())
+                        }),
+                        concrete_path,
+                        alloc_tgt.tag(),
+                        &alloc_tag,
+                    )?;
+                }
+                None => (),
+            }
+        }
+
+        // Create hash enforce
+        for (slot, tgt, tag) in enforce_data {
+            // get alloc_tgt from tgt
+            let Some(alloc_tgt) = alloc_ptrs.get(tgt.name()) else {
+                bail!("{} not allocated", tgt.name());
+            };
+
+            // get slot_hash from slot name
+            let Some(slot_hash) = hash_slots.get(slot) else {
+                bail!("Slot number {} not allocated", slot)
+            };
+
+            enforce_equal(
+                cs,
+                || {
+                    format!(
+                        "enforce equal hash for tgt {} and slot number {}",
+                        tgt.name(),
+                        slot,
+                    )
+                },
+                alloc_tgt.hash(),
+                slot_hash,
+            );
+
+            if let Some(tag) = tag {
```

Comment:
Why is the tag optional? Under what circumstances would it be absent?

Reply:
UnHash operations don't have a tag, at least not yet. A LEM programmer can use MatchTag to do whatever is necessary depending on the tag value. But this point was something that we discussed a bit already, and we were not sure the best way to proceed, so we basically kept the old interface (without tag). If we add a `tag` parameter to `UnHash` operations, then implementation would be symmetric. We use `UnHash` for destructuring expressions, for strings, etc. So indeed a tag validation must happen somewhere.

Reply:
We've made the change to make it symmetric